torch/nn/functional.py

```diff
@@ -2316,6 +2316,65 @@ def poisson_nll_loss(input, target, log_input=True, full=False, size_average=Non
     return ret
 
 
+def gaussian_nll_loss(input, target, var, *, eps=1e-6, full=False, reduction='mean'):
+    r"""Gaussian negative log likelihood loss.
+
+    See :class:`~torch.nn.GaussianNLLLoss` for details.
+
+    Args:
+        input: expectation of the Gaussian distribution.
+        target: sample from the Gaussian distribution.
+        var: tensor of positive variance(s), one for each of the expectations
+            in the input (heteroscedastic), or a single one (homoscedastic).
+        eps: value added to var, for stability. Default: 1e-6.
+        full: ``True``/``False`` (bool), include the constant term in the loss
+            calculation. Default: ``False``.
+        reduction: specifies the reduction to apply to the output:
+            ``'none'`` | ``'mean'`` | ``'sum'``. ``'none'``: no reduction will be applied,
+            ``'mean'``: the output is the average of all batch member losses,
+            ``'sum'``: the output is the sum of all batch member losses.
+            Default: ``'mean'``.
+    """
+    # Inputs and targets much have same shape
+    input = input.view(input.size(0), -1)
+    target = target.view(target.size(0), -1)
+    if input.size() != target.size():
+        raise ValueError("input and target must have same size")
+
+    # Second dim must be 1 ie shape (N, 1) if var and input do not have same shape
+    var = var.view(var.size(0), -1)
+    if var.size() != input.size() and var.size(1) != 1:
+        raise ValueError("var is of incorrect size")
+
+    if reduction != 'none' and reduction != 'mean' and reduction != 'sum':
+        raise ValueError(reduction + " is not valid")
+
+    # Entries of var must be non-negative
+    if torch.any(var < 0):
+        raise ValueError("var has negative entry/entries")
+
+    # Clamp for stability
+    var = var.clone()
+    with torch.no_grad():
+        var.clamp_(min=eps)
```

Comment:
@ezyang  I suggested here that we use these 3 lines instead of doing `var = var + eps` here to reduce the impact of the epsilon for points that are far from 0.
Does that look like a good idea to you? Or you think we shouldn't do that?

Reply:
I don't have an opinion. With these sorts of stability considerations I'd prefer that we do what is usually done in the literature, but I'm not widely read enough to know what it is here.

Reply:
My opinion here is that I agree with @albanD's suggestion. While it is different than what e.g. `PoissonNLLLoss` does, it matches the literature.  

Quoting the [paper](https://arxiv.org/pdf/1612.01474.pdf): "We enforce the positivity constraint on the variance by passing the second output through the softplus
function, and add a minimum variance (e.g. 10^-6) for numerical stability."

The phrase "minimum variance" suggests to me that the clamping behavior is desired over an unconditional addition of `eps`.

Reply:
Thanks for checking @jbschlosser !
So that's good and we just should update the formula to match this.

Reply:
There was a similar discussion about cosine similarity a year ago, went nowhere https://github.com/pytorch/pytorch/pull/31378#discussion_r362756119. 
cdist function that also has problem with occasional zero values uses clamping. 
BatchNorm and other normalizations use var + eps, in accordance with their original papers.

Reply:
Ho I forgot about this one :D 
Thanks for the link! I guess that following the paper is the right thing to do right now.

Reply:
I would still like us to have a consistent story for norms and this function (and cosine similarity for that matter), but it shouldn't block this PR.

Reply:
@almson See this discussion of `var` clamping